File: Java Based Archives/RetrofitDemo/app/src/main/java/com/android/retrofitdemo/MainActivity.java
package com.android.retrofitdemo;

import android.os.Bundle;
import android.util.Log;
import androidx.appcompat.app.AppCompatActivity;
import com.android.retrofitdemo.model.Data;
import com.android.retrofitdemo.service.RestCountriesService;
import com.android.retrofitdemo.service.RetroFitInstance;
import java.util.List;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Call the method to fetch data
        getData();
    }

    private void getData() {
        // Create an instance of the RestCountriesService using RetrofitInstance
        RestCountriesService service = RetroFitInstance.getRestCountriesService();

        // Make the API call using the service
        Call<List<Data>> call = service.getAllCountries();

        // Execute the call asynchronously
        call.enqueue(new Callback<List<Data>>() {
            @Override
            public void onResponse(Call<List<Data>> call, Response<List<Data>> response) {
                if (response.isSuccessful()) {
                    // Handle successful response
                    List<Data> dataList = response.body();
                    if (dataList != null && !dataList.isEmpty()) {
                        // Data retrieval successful, handle the data here
                        processData(dataList);
                    } else {
                        Log.e("MainActivity", "No data received");
                    }
                } else {
                    // Handle unsuccessful response
                    Log.e("MainActivity", "Failed to fetch data: " + response.code());
                }
            }

            @Override
            public void onFailure(Call<List<Data>> call, Throwable t) {
                // Handle failure
                Log.e("MainActivity", "Error fetching data: " + t.getMessage());
            }
        });
    }
    private void processData(List<Data> dataList) {
        // Process and display data as needed
        for (Data data : dataList) {
            Log.d("MainActivity", "Country Name: " + data.getName());

            // Check if the capital list is not null and not empty before accessing its elements
            List<String> capitalList = data.getCapital();
            if (capitalList != null && !capitalList.isEmpty()) {
                Log.d("MainActivity", "Capital: " + capitalList.get(0));
            } else {
                Log.e("MainActivity", "Capital list is null or empty for country: " + data.getName());
            }

            // Log additional data or perform other operations
        }
    }



}

File: RetrofitDemo/app/src/main/java/com/anushka/retrofitdemo/MainActivity.kt
package com.anushka.retrofitdemo

import android.os.Bundle
import android.util.Log
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.liveData
import retrofit2.Response

class MainActivity : AppCompatActivity() {

    // Retrofit service instance for making API calls
    private val retService: AlbumService = RetrofitInstance
        .getRetrofitInstance()
        .create(AlbumService::class.java)

    // TextView for displaying API response in the UI
    private lateinit var textView: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Initialize the TextView
        textView = findViewById(R.id.textView)

        // Make API calls and observe responses
        getRequestWithQueryParameter()
        getRequestWithPathParameter()
    }

    // Function to make an API call with query parameters
    private fun getRequestWithQueryParameter() {
        // Create a LiveData to handle the response from the API call
        val responseLiveData: LiveData<Response<Album>> = liveData {
            try {
                // Make the API call and emit the response
                val response: Response<Album> = retService.getAlbums()
                emit(response)
            } catch (e: Exception) {
                // Handle any exceptions that might occur during the API call
                Log.e("MyTag", "Error: ${e.message}")
            }
        }

        // Observe the LiveData for changes in the API response
        observeAlbumResponse(responseLiveData)
    }

    // Function to make an API call with a path parameter
    private fun getRequestWithPathParameter() {
        // Create a LiveData to handle the response from the API call
        val pathResponse: LiveData<Response<AlbumItem>> = liveData {
            // Make the API call with a path parameter and emit the response
            val response: Response<AlbumItem> = retService.getAlbum(3)
            emit(response)
        }

        // Observe the LiveData for changes in the API response
        observePathResponse(pathResponse)
    }

    // Function to observe and handle the response from the API with query parameters
    private fun observeAlbumResponse(responseLiveData: LiveData<Response<Album>>) {
        responseLiveData.observe(this, Observer {
            // Check if the API response is successful
            if (it.isSuccessful) {
                // Extract the list of albums from the response body
                val albumsList = it.body()?.listIterator()
                albumsList?.forEach { albumItem ->
                    // Display album information in the TextView
                    val result: String = " " + "Album Title : ${albumItem.title}" + "\n" +
                            " " + "Album id : ${albumItem.id}" + "\n" +
                            " " + "User id : ${albumItem.userId}" + "\n" + "\n\n\n"
                    textView.append(result)
                    // Log album title to the console
                    Log.i("MyTag", albumItem.title)
                }
            } else {
                // Log an error message if the API response is not successful
                Log.e("MyTag", "API call failed with code: ${it.code()}")
            }
        })
    }

    // Function to observe and handle the response from the API with path parameters
    private fun observePathResponse(pathResponse: LiveData<Response<AlbumItem>>) {
        pathResponse.observe(this, Observer {
            // Extract the title from the response body
            val title: String? = it.body()?.title
            // Display a Toast message with the title
            Toast.makeText(applicationContext, title, Toast.LENGTH_LONG).show()
        })
    }
}
